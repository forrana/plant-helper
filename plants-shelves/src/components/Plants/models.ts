export interface PlantType {
  id: number;
  name: string;
  scientificName: string;
  daysUntilNextWatering: number;
  daysBetweenWatering: number;
  daysBetweenWateringGrowing: number;
  daysBetweenWateringDormant: number;
  symbol: SymbolType;
  daysPostpone: number;
}

export interface SymbolType {
  userWideId: number
}

export interface PlantsData {
  plants: PlantType[];
}

export interface Subscription {
  endpoint: string,
  p256dh: string,
  auth: string,
  permission_given: boolean
}

export interface SubscriptionData {
  subscription: boolean
}

export interface PlantData {
  plant: PlantType;
}

export interface PlantSuggestion {
  nickName: string,
  nickNames: Array<string>,
  scientificName: string,
  daysBetweenWateringGrowingInt: number,
  daysBetweenWateringDormantInt: number
}

export interface PlantNickName {
  name: string,
  plantEntry: PlantSuggestion
}

export interface GlobalState extends PlantsData { }

export interface loadAction { type: 'load', plants: PlantType[] }
export interface addAction { type: 'add', plant: PlantType }
export interface updateAction { type: 'update', plant: PlantType, index: number }
export interface deleteAction { type: 'delete', index: number }

export type GlobalReducerAction =
  | addAction
  | loadAction
  | updateAction
  | deleteAction;
