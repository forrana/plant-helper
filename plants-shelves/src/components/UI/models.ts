export interface MessageType { description: string, color: "danger" | "info" | "warning" | "success" }

export interface AddMessage  { type: 'addMessage', message: MessageType }
export interface RemoveMessage  { type: 'removeMessage' }

export interface AlertState { messages: MessageType[] }

export type AlertReducerAction =
  | AddMessage
  | RemoveMessage