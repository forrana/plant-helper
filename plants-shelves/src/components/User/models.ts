export interface LoginError {
  message: string;
  code: string;
}
