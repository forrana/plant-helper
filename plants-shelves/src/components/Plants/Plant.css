.plant {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  border: 1px solid black;
  margin: 0.25em;
  padding: 0.25em;
}
