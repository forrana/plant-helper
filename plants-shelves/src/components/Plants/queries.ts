import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
    symbol {
      userWideId
    }
    room {
      id,
      roomName,
      colorBackground
    }
  }
}
`;

const GET_ALL_ROOMS = gql`
  query {
    rooms {
      id, roomName, colorBackground
    }
  }
`;
const ROOMS_BY_NAME_FRAGMENT = gql`
  query RoomsByNameFragment($nameFragment: String!) {
    roomsByNameFragment(nameFragment: $nameFragment) {
      id, roomName, colorBackground
    }
  }
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $groupName: String, $colorBackground: String) {
      createPlant(plantName: $plantName, scientificName: $scientificName, daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
        daysBetweenWateringDormant: $daysBetweenWateringDormant, groupName: $groupName, colorBackground: $colorBackground) {
        plant {
          id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
          symbol {
            userWideId
          }
          room {
            id,
            roomName,
            colorBackground
          }
        }
      }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const POSTPONE_WATERING = gql`
  mutation PostponeWatering($plantId: ID!, $postponeDays: Int!) {
    postponeWatering(plantId: $plantId, days: $postponeDays) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_PLANT = gql`
  mutation UpdatePlant($plantId: ID!, $plantName: String!,
    $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $postponeDays: Int!, $groupName: String, $colorBackground: String) {
    updatePlant(plantId: $plantId, plantName: $plantName, scientificName: $scientificName,
      daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
      daysBetweenWateringDormant: $daysBetweenWateringDormant,
      postponeDays: $postponeDays,
      groupName: $groupName,
      colorBackground: $colorBackground) {
      plant {
        id, name, scientificName, daysUntilNextWatering,
        daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant,
        daysPostpone,
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_ROOM = gql`
  mutation UpdateRoom($roomId: ID!, $roomName: String!) {
    updateRoom(roomId: $roomId, roomName: $roomName) {
      ok
      room {
        id,
        roomName,
        colorBackground
      }
    }
  }
`

const CREATE_ROOM = gql`
  mutation CreateRoom($plant1Id: ID!, $plant2Id: ID!) {
    createRoom(plant1Id: $plant1Id, plant2Id: $plant2Id) {
      ok
    }
  }
`;

const DELETE_PLANT = gql`
  mutation DeletePlant($plantId: ID!) {
    deletePlant(plantId: $plantId) {
      ok
    }
  }
`;

const GET_SUBSCRIPTION = gql`
  query {
    subscription
  }
`

const CREATE_SUBSCRIPTION = gql`
  mutation CreateSubscription($endpoint: String!, $p256dh: String!, $auth: String!, $permissionGiven: Boolean!) {
    createSubscription(endpoint: $endpoint, p256dh: $p256dh, auth: $auth, permissionGiven: $permissionGiven) {
      ok
    }
  }
`

export {
  GET_ALL_PLANTS,
  GET_ALL_ROOMS,
  ADD_PLANT,
  WATER_PLANT,
  UPDATE_PLANT,
  DELETE_PLANT,
  CREATE_SUBSCRIPTION,
  GET_SUBSCRIPTION,
  POSTPONE_WATERING,
  CREATE_ROOM,
  ROOMS_BY_NAME_FRAGMENT,
  UPDATE_ROOM
}
