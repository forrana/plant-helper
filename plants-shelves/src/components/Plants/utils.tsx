const getRandomString = (string_length: number): string => [...Array(string_length)].map(i=>(~~(Math.random()*36)).toString(36)).join('')

export { getRandomString }