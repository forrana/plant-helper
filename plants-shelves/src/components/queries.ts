import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
  }
}
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!) {
    createPlant(plantName: $plantName, scientificName: $scientificName) {
      plant {
        name,
        scientificName
      }
    }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
      }
    }
  }
`;


export { GET_ALL_PLANTS, ADD_PLANT, WATER_PLANT }
