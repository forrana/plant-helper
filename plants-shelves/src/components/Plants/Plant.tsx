import React, { useContext, useState } from 'react';
import { useMutation } from '@apollo/client';
import { PlantData } from './models'
import { Card, CardBody, CardTitle, CardSubtitle, Progress, Button, Spinner,
         Modal, ModalHeader, ModalBody, ModalFooter, Badge,
        } from 'reactstrap';
import { WATER_PLANT, DELETE_PLANT } from './queries'
import PlantsDispatch from './PlantsDispatch'
import PlantsEdit from './PlantsEdit'
import styles from "./Plant.module.css"
import uiStyles from "../UI/UIElements.module.css"
import ErrorHandler from './ErrorHandler';

interface PlantProps extends PlantData { index: number }
interface WhenToWaterProps { daysUntilNextWatering: number }


function WhenToWater({ daysUntilNextWatering }: WhenToWaterProps) {
  if(daysUntilNextWatering > 1)
    return (
      <div className="text-center">{daysUntilNextWatering} days until watering</div>
    )

  if(daysUntilNextWatering === 1)
    return (
      <div className="text-center">Water in {daysUntilNextWatering} day</div>
    )

  if(daysUntilNextWatering < 1)
    return (
      <div className="text-center">Water me!</div>
    )

  return (
    <div className="text-center">Uups! Something went wrong!</div>
  )
}

function Plant({ plant, index }: PlantProps) {
    const dispatch = useContext(PlantsDispatch);

    const [isEditMode, setIsEditMode] = useState(false);
    const toggleEditMode = () => setIsEditMode(!isEditMode);

    const [modal, setModal] = useState(false);
    const toggleModal = () => setModal(!modal);

    const [waterPlant, wateringStatus] = useMutation(WATER_PLANT, {
      onCompleted: (data: { waterPlant: PlantData }) => {
        dispatch && dispatch({ type: 'update', plant: data.waterPlant.plant, index: index })
      },
      onError: (e) => console.error('Error creating plant:', e)
    });

    const toWater = () => waterPlant({variables: { plantId: plant.id } });

    const [deletePlant, deletingStatus] = useMutation(DELETE_PLANT, {
      onCompleted: () => {
        dispatch && dispatch({ type: 'delete', index: index })
      },
      onError: (e) => console.error('Error creating plant:', e)
    });

    const confirmDeletion = () => {
      deletePlant({variables: { plantId: plant.id } });
      toggleModal()
    }
    const plantHealth = (plant.daysUntilNextWatering / plant.daysBetweenWatering)*100;
    // TODO should be memoized perhaps?
    const healthColor = () => {
      if( plantHealth >= 70) {
          return "success"
        } else if ( plantHealth >= 50 ) {
          return "info"
        } else if ( plantHealth > 20 ) {
          return "warning"
        } else {
          return "danger"
        }
    }

    const backgroundColor = () => {
      if( plantHealth >= 70) {
          return "success"
        } else if ( plantHealth >= 50 ) {
          return "info"
        } else if ( plantHealth > 20 ) {
          return "warning"
        } else if ( plantHealth > 5 ) {
          return "danger"
        } else {
          return "alarm"
        }
    }

    if (wateringStatus.loading || deletingStatus.loading) return  <Spinner color="primary" />

    if (wateringStatus.error) {
      return  <ErrorHandler error={wateringStatus.error} />
    }

    if (deletingStatus.error) {
      return  <ErrorHandler error={deletingStatus.error} />
    }

    if (isEditMode)
      return (
        <Card className={styles.plant}>
          <Badge color="light">{plant.id}</Badge>
          <CardBody>
            <PlantsEdit plant={plant} index={index} action={toggleEditMode}/>
          </CardBody>
        </Card>
      )

    return (
      <Card className={`${styles.plant} ${styles[backgroundColor()]}`} data-testid={`plant-card-${index}`}>
        <Badge color="light">{plant.id}</Badge>
        <CardBody>
          <section className={styles.controls} data-testid={`plant-controls-${index}`}>
            <Button outline onClick={toWater} title="Water" data-testid="water-btn" className={uiStyles.roundButton}>&#128166;</Button>
            <Button outline onClick={toggleEditMode} title="Edit" data-testid="edit-btn" className={uiStyles.roundButton}>&#x270E;</Button>
            <Button outline onClick={toggleModal} title="Remove" data-testid="remove-btn" className={uiStyles.roundButton}>&#x1F5D1;</Button>
          </section>
          <div className={styles.image}>&#129716;</div>
          <CardTitle tag="h5">{plant.name}</CardTitle>
          <CardSubtitle tag="h6" className="mb-2 text-muted">{plant.scientificName}</CardSubtitle>
          <WhenToWater daysUntilNextWatering={plant.daysUntilNextWatering}/>
          <Progress value={ plantHealth } color={ healthColor() }/>
        </CardBody>

        <Modal isOpen={modal} toggle={toggleModal}>
          <ModalHeader toggle={toggleModal}>Delete plant {plant.name} </ModalHeader>
          <ModalBody>
            Warning! This action is irreversible.
          </ModalBody>
          <ModalFooter className={uiStyles.footer}>
            <Button color="danger" onClick={confirmDeletion} data-testid="modal-button-delete">Delete!</Button>
            <Button color="secondary" onClick={toggleModal}>Cancel!</Button>
          </ModalFooter>
        </Modal>

      </Card>
    )
}

export { Plant }
