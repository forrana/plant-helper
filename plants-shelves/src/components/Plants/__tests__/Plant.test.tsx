import React from 'react';
import { act, fireEvent, render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import { GlobalState, PlantType } from '../models';
import { DELETE_PLANT, GET_ALL_PLANTS, WATER_PLANT } from '../queries';
import { Plant } from '../Plant';
import { getRandomString } from '../utils'

const getPlantWithXDaysToWater = (days: number): PlantType => ({
  id: 1,
  name: "Aloe 1",
  scientificName: "Aloe Vera 1" ,
  daysUntilNextWatering: days,
  daysBetweenWatering: 7,
  symbol: {
    userWideId: 1
  }
})

const plantUnWatered: PlantType = getPlantWithXDaysToWater(0)
const plant1DayToWater: PlantType = getPlantWithXDaysToWater(1)
const plant4DaysToWater: PlantType = getPlantWithXDaysToWater(4)
const plant2DaysToWater: PlantType = getPlantWithXDaysToWater(2)
const plantWIthErrornousId: PlantType = {
  id: 100,
  name: "Aloe 1",
  scientificName: "Aloe Vera 1" ,
  daysUntilNextWatering: 7,
  daysBetweenWatering: 7,
  symbol: {
    userWideId: 1
  }
}

const errorMessage: string = `error-${getRandomString(10)}`

const plantWatered: PlantType = {
  ...plantUnWatered, daysUntilNextWatering: 7
};
const mocksWithPlant: any = [
  {
    request: {
      query: GET_ALL_PLANTS,
    },
    result: {
      data: {
        plants: [plantUnWatered, plant1DayToWater],
      },
    },
  },
  {
    request: {
      query: WATER_PLANT,
      variables: { plantId: plantUnWatered.id }
    },
    result: {
      data:  plantWatered,
    },
  },
  {
    request: {
      query: DELETE_PLANT,
      variables: { plantId: plantUnWatered.id }
    },
    result: {
      data: { ok: true },
    },
  },
  {
    request: {
      query: WATER_PLANT,
      variables: { plantId: plantWIthErrornousId.id }
    },
    error: {
      message: errorMessage
    },
  },
  {
    request: {
      query: DELETE_PLANT,
      variables: { plantId: plantWIthErrornousId.id }
    },
    error: {
      message: errorMessage
    },
  },
];

test('Should indicate water me', async () => {
  const {container} = render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plantUnWatered} index={0}/>
    </MockedProvider>,
  );

  expect(container).toMatchSnapshot()
})

test('Should indicate 1 day to water', async () => {
  const {container} = render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plant1DayToWater} index={0}/>
    </MockedProvider>,
  );

  expect(container).toMatchSnapshot()
})

test('Should indicate 4 days to water', async () => {
  const {container} = render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plant4DaysToWater} index={0}/>
    </MockedProvider>,
  );

  expect(container).toMatchSnapshot()
})

test('Should indicate 2 days to water', async () => {
  const {container} = render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plant2DaysToWater} index={0}/>
    </MockedProvider>,
  );

  expect(container).toMatchSnapshot()
})

test('Should ask for watering', async () => {
  render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plantUnWatered} index={0}/>
    </MockedProvider>,
  );

  await act(async () => {
    fireEvent(
      screen.getByTitle('Water'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
    await new Promise(resolve => setTimeout(resolve, 0));
  });

  const wateringText = screen.getByText(/Water me!/);
  expect(wateringText).toBeInTheDocument();
});

test('Watering button to be present', async () => {
  render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plantUnWatered} index={0}/>
    </MockedProvider>,
  );
  const wateringButton = screen.getByTitle('Water');
  expect(wateringButton).toBeInTheDocument();
});


test('Be able to enter into edit mode', async () => {
  render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plantUnWatered} index={0}/>
    </MockedProvider>,
  );

  await act(async () => {
    fireEvent(
      screen.getByTitle('Edit'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  });

  const saveButton = screen.getByTitle(/Save!/);
  expect(saveButton).toBeInTheDocument();
});

test('Show confirmation modal on delete action', async () => {
  render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plantUnWatered} index={0}/>
    </MockedProvider>,
  );

  await act(async () => {
    fireEvent(
      screen.getByTitle('Remove'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  });

  const deleteButton = screen.getByText(/Delete!/);
  expect(deleteButton).toBeVisible()
});

test('Hide confirmation modal after deletion', async () => {
  render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plantUnWatered} index={0}/>
    </MockedProvider>,
  );

  await act(async () => {
    fireEvent(
      screen.getByTitle('Remove'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  });

  const deleteButton = screen.getByText(/Delete!/);
  expect(deleteButton).toBeVisible()

  await act(async () => {
    fireEvent(
      deleteButton,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  });

  expect(deleteButton).not.toBeVisible()
});

test('Indicate deletion error', async () => {
  render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plantWIthErrornousId} index={0}/>
    </MockedProvider>,
  );

  await act(async () => {
    fireEvent(
      screen.getByTitle('Remove'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  });

  await act(async () => {
    fireEvent(
      screen.getByText(/Delete!/),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  });

  const message = screen.getByText(new RegExp(errorMessage, "i"));
  expect(message).toBeVisible()
});

test('Indicate watering error', async () => {
  render(
    <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <Plant plant={plantWIthErrornousId} index={0}/>
    </MockedProvider>,
  );

  await act(async () => {
    fireEvent(
      screen.getByTitle('Water'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  });

  const message = screen.getByText(new RegExp(errorMessage, "i"));
  expect(message).toBeVisible()

});
