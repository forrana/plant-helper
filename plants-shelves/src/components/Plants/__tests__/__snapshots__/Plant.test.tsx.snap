// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should indicate 1 day to water 1`] = `
<div>
  <div
    class="plant danger card"
    data-testid="plant-card-0"
  >
    <span
      class="badge badge-light"
    >
      1
    </span>
    <div
      class="card-body"
    >
      <section
        class="controls"
        data-testid="plant-controls-0"
      >
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="water-btn"
          title="Water"
          type="button"
        >
          ðŸ’¦
        </button>
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="edit-btn"
          title="Edit"
          type="button"
        >
          âœŽ
        </button>
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="remove-btn"
          title="Remove"
          type="button"
        >
          ðŸ—‘
        </button>
      </section>
      <div
        class="image"
      >
        ðŸª´
      </div>
      <h5
        class="card-title"
      >
        Aloe 1
      </h5>
      <h6
        class="mb-2 text-muted card-subtitle"
      >
        Aloe Vera 1
      </h6>
      <div
        class="text-center"
      >
        Water in 
        1
         day
      </div>
      <div
        class="progress"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="14.285714285714285"
          class="progress-bar bg-danger"
          role="progressbar"
          style="width: 14.285714285714285%;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Should indicate 2 days to water 1`] = `
<div>
  <div
    class="plant warning card"
    data-testid="plant-card-0"
  >
    <span
      class="badge badge-light"
    >
      1
    </span>
    <div
      class="card-body"
    >
      <section
        class="controls"
        data-testid="plant-controls-0"
      >
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="water-btn"
          title="Water"
          type="button"
        >
          ðŸ’¦
        </button>
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="edit-btn"
          title="Edit"
          type="button"
        >
          âœŽ
        </button>
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="remove-btn"
          title="Remove"
          type="button"
        >
          ðŸ—‘
        </button>
      </section>
      <div
        class="image"
      >
        ðŸª´
      </div>
      <h5
        class="card-title"
      >
        Aloe 1
      </h5>
      <h6
        class="mb-2 text-muted card-subtitle"
      >
        Aloe Vera 1
      </h6>
      <div
        class="text-center"
      >
        2
         days until watering
      </div>
      <div
        class="progress"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="28.57142857142857"
          class="progress-bar bg-warning"
          role="progressbar"
          style="width: 28.57142857142857%;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Should indicate 4 days to water 1`] = `
<div>
  <div
    class="plant info card"
    data-testid="plant-card-0"
  >
    <span
      class="badge badge-light"
    >
      1
    </span>
    <div
      class="card-body"
    >
      <section
        class="controls"
        data-testid="plant-controls-0"
      >
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="water-btn"
          title="Water"
          type="button"
        >
          ðŸ’¦
        </button>
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="edit-btn"
          title="Edit"
          type="button"
        >
          âœŽ
        </button>
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="remove-btn"
          title="Remove"
          type="button"
        >
          ðŸ—‘
        </button>
      </section>
      <div
        class="image"
      >
        ðŸª´
      </div>
      <h5
        class="card-title"
      >
        Aloe 1
      </h5>
      <h6
        class="mb-2 text-muted card-subtitle"
      >
        Aloe Vera 1
      </h6>
      <div
        class="text-center"
      >
        4
         days until watering
      </div>
      <div
        class="progress"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="57.14285714285714"
          class="progress-bar bg-info"
          role="progressbar"
          style="width: 57.14285714285714%;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Should indicate water me 1`] = `
<div>
  <div
    class="plant alarm card"
    data-testid="plant-card-0"
  >
    <span
      class="badge badge-light"
    >
      1
    </span>
    <div
      class="card-body"
    >
      <section
        class="controls"
        data-testid="plant-controls-0"
      >
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="water-btn"
          title="Water"
          type="button"
        >
          ðŸ’¦
        </button>
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="edit-btn"
          title="Edit"
          type="button"
        >
          âœŽ
        </button>
        <button
          class="roundButton btn btn-outline-secondary"
          data-testid="remove-btn"
          title="Remove"
          type="button"
        >
          ðŸ—‘
        </button>
      </section>
      <div
        class="image"
      >
        ðŸª´
      </div>
      <h5
        class="card-title"
      >
        Aloe 1
      </h5>
      <h6
        class="mb-2 text-muted card-subtitle"
      >
        Aloe Vera 1
      </h6>
      <div
        class="text-center"
      >
        Water me!
      </div>
      <div
        class="progress"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="0"
          class="progress-bar bg-danger"
          role="progressbar"
          style="width: 0%;"
        />
      </div>
    </div>
  </div>
</div>
`;
